app/controllers/api/v0/api_controller.rb

```diff
@@ -29,6 +29,52 @@ def auth_results
     render json: { status: "ok" }
   end
 
+  def user_qualification_data
+    user = User.find(params.require(:user_id))
+    begin
+      cutoff_date = date_from_params || Date.current
+    rescue Date::Error
+      return render json: { error: 'Invalid date format. Please provide an iso8601 date string.' }, status: :bad_request
+    end
+
+    return render json: { error: 'You cannot request qualification data for a future date.' }, status: :bad_request if cutoff_date > Date.current
+    cutoff_date = cutoff_date.iso8601
+
+    qualification_results = []
+
+    # Compile singles
+    best_singles_by_cutoff = user.person&.best_singles_by(cutoff_date)
+    return render json: [] unless best_singles_by_cutoff.present? # If the user has no best singles, they have no qualification-relevant results
+    best_singles_by_cutoff.map do |event, time|
+      qualification_results << {
+        eventId: event,
+        type: 'single',
+        best: time,
+        on_or_before: cutoff_date,
+      }
+    end
+
+    # Compile averages
+    best_averages_by_cutoff = user.person&.best_averages_by(cutoff_date)
+    return render json: qualification_results unless best_averages_by_cutoff.present? # If the user has no best averages, we can just return their singles
+    best_averages_by_cutoff.map do |event, time|
```

Comment:
have a function that takes event, type, time and cutoff as parameteres to return the hash I'm building here